File: python/dsa/algorithms/sliding-window/sliding-window-maximum.py
def max_sliding_window(nums: List[int], k: int) -> List[int]:
    q = deque()
    res = []
    left = 0

    for right in range(len(nums)):
        # 1️⃣ Expand: Maintain monotonic property
        while q and nums[q[-1]] <= nums[right]:
            q.pop()
        q.append(right);

        // 2️⃣ When window size reaches k
        if right - left + 1 == k:
            // 3️⃣ Record / Evaluate
            res.append(nums[q[0]])

            // 4️⃣ Slide window
            if q[0] == left:
                q.popleft()
            left += 1

    return res

File: javascript/dsa/algorithms/sliding-window/sliding-window-maximum.js
function maxSlidingWindow(nums, k) {
    let q = [];
    let res = [];
    let left = 0;

    for (let right = 0; right < nums.length; right++) {
        // 1️⃣ Expand: Maintain monotonic property
        while (q.length && nums[q[q.length - 1]] <= nums[right]) {
            q.pop();
        }
        q.push(right);

        // 2️⃣ When window size reaches k
        if (right - left + 1 === k) {
            // 3️⃣ Record / Evaluate
            res.push(nums[q[0]]);
        }
    }

    return res;
}